File: server/lib/env_config.py
"""
Centralized environment configuration for enabling/disabling deployment-specific behavior.

This module provides environment-aware toggles that allow temporarily disabling
production/deployment behavior while preserving the existing production architecture.
"""

import os
from typing import List


def _get_bool_env(var_name: str, default: bool = False) -> bool:
    """Get boolean environment variable with default."""
    value = os.environ.get(var_name, "")
    if not value:
        return default
    return value.lower() in ("true", "1", "yes", "on")


def _get_app_env() -> str:
    """Get application environment."""
    return os.environ.get("APP_ENV", "development")


# Application environment
APP_ENV = _get_app_env()

# Environment flags for controlling deployment-specific behavior
USE_LOCAL_STORAGE = _get_bool_env("USE_LOCAL_STORAGE", APP_ENV == "development")
USE_SUPABASE_STORAGE = _get_bool_env("USE_SUPABASE_STORAGE", APP_ENV == "production")
USE_PRODUCTION_COOKIES = _get_bool_env("USE_PRODUCTION_COOKIES", APP_ENV == "production")
USE_PRODUCTION_URLS = _get_bool_env("USE_PRODUCTION_URLS", APP_ENV == "production")
USE_RAILWAY_DETECTION = _get_bool_env("USE_RAILWAY_DETECTION", APP_ENV == "production")
USE_VERCEL_DETECTION = _get_bool_env("USE_VERCEL_DETECTION", APP_ENV == "production")

# Development/production flags
IS_DEVELOPMENT = APP_ENV == "development"
IS_PRODUCTION = APP_ENV == "production"


def get_api_base_url() -> str:
    """
    Get the appropriate API base URL based on environment flags.
    
    In development: http://127.0.0.1:5000/api
    In production: Uses API_BASE_URL or falls back to Railway detection
    """
    # If explicitly using production URLs, use the configured API base URL
    if USE_PRODUCTION_URLS:
        explicit = os.environ.get("API_BASE_URL", "").strip()
        if explicit:
            return explicit.rstrip("/")
        
        # Fall back to Railway detection if enabled
        if USE_RAILWAY_DETECTION:
            railway = os.environ.get("RAILWAY_PUBLIC_DOMAIN", "").strip()
            if railway:
                return f"https://{railway}/api"
    
    # Default to localhost API for development
    return "http://127.0.0.1:5000/api"


def get_frontend_url() -> str:
    """
    Derive canonical frontend URL from env or build from Railway/Vercel.
    """
    explicit = os.environ.get("FRONTEND_URL", "").strip()
    if explicit:
        return explicit.rstrip("/")
    
    # Use production URLs if enabled
    if USE_PRODUCTION_URLS:
        vercel = os.environ.get("VERCEL_URL", "").strip()
        if vercel and USE_VERCEL_DETECTION:
            return f"https://{vercel}"
        
        vercel_branch = os.environ.get("VERCEL_BRANCH_URL", "").strip()
        if vercel_branch and USE_VERCEL_DETECTION:
            return f"https://{vercel_branch}"
        
        railway = os.environ.get("RAILWAY_PUBLIC_DOMAIN", "").strip()
        if railway and USE_RAILWAY_DETECTION:
            return f"https://{railway}"
    
    # Default to localhost for development
    return "http://localhost:3000"


def get_cors_origins(frontend_url: str) -> List[str]:
    """
    Build allowed CORS origins from FRONTEND_URL, CORS_ORIGINS env, and defaults.
    """
    LOCAL_FALLBACKS = ("http://localhost:3000", "http://127.0.0.1:3000")

    raw = os.environ.get("CORS_ORIGINS", "").strip()
    origins: List[str] = []
    seen: set[str] = set()

    for part in raw.replace(";", ",").split(","):
        origin = part.strip().strip('"').strip("'")
        if origin and origin not in seen:
            seen.add(origin)
            origins.append(origin)

    if frontend_url and frontend_url not in seen:
        seen.add(frontend_url)
        origins.append(frontend_url)

    # Add local fallbacks in development
    if IS_DEVELOPMENT or frontend_url == "http://localhost:3000":
        for fallback in LOCAL_FALLBACKS:
            if fallback not in seen:
                seen.add(fallback)
                origins.append(fallback)

    # Add Railway domain if enabled
    if USE_RAILWAY_DETECTION:
        railwy_domain = os.environ.get("RAILWAY_PUBLIC_DOMAIN", "").strip()
        if railwy_domain:
            railwy_origin = f"https://{railwy_domain}"
            if railwy_origin not in seen:
                seen.add(railwy_origin)
                origins.append(railwy_origin)

    return origins


class EnvFlags:
    """Environment flags for controlling deployment-specific behavior."""
    
    APP_ENV = APP_ENV
    USE_LOCAL_STORAGE = USE_LOCAL_STORAGE
    USE_SUPABASE_STORAGE = USE_SUPABASE_STORAGE
    USE_PRODUCTION_COOKIES = USE_PRODUCTION_COOKIES
    USE_PRODUCTION_URLS = USE_PRODUCTION_URLS
    USE_RAILWAY_DETECTION = USE_RAILWAY_DETECTION
    USE_VERCEL_DETECTION = USE_VERCEL_DETECTION
    IS_DEVELOPMENT = IS_DEVELOPMENT
    IS_PRODUCTION = IS_PRODUCTION
    
    @staticmethod
    def get_api_base_url() -> str:
        return get_api_base_url()
    
    @staticmethod
    def get_frontend_url() -> str:
        return get_frontend_url()
    
    @staticmethod
    def get_cors_origins(frontend_url: str) -> List[str]:
        return get_cors_origins(frontend_url)

File: client/lib/env-config.ts
/**
 * Centralized environment configuration for enabling/disabling deployment-specific behavior.
 * 
 * This module provides environment-aware toggles that allow temporarily disabling
 * production/deployment behavior while preserving the existing production architecture.
 * 
 * Usage:
 *   import { EnvFlags } from "@/lib/env-config";
 *   
 *   if (EnvFlags.USE_LOCAL_STORAGE) {
 *     // use local storage
 *   } else {
 *     // use Supabase storage (production)
 *   }
 */

// Read environment variables with defaults for development
const APP_ENV = process.env.APP_ENV || "development";
const USE_LOCAL_STORAGE = 
  process.env.USE_LOCAL_STORAGE?.toLowerCase() === "true" || 
  APP_ENV === "development";

const USE_SUPABASE_STORAGE = 
  process.env.USE_SUPABASE_STORAGE?.toLowerCase() === "true" && 
  APP_ENV !== "development";

const USE_PRODUCTION_COOKIES = 
  process.env.USE_PRODUCTION_COOKIES?.toLowerCase() === "true" && 
  APP_ENV === "production";

const USE_PRODUCTION_URLS = 
  process.env.USE_PRODUCTION_URLS?.toLowerCase() === "true" && 
  APP_ENV === "production";

const USE_RAILWAY_DETECTION = 
  process.env.USE_RAILWAY_DETECTION?.toLowerCase() === "true" && 
  APP_ENV === "production";

const USE_VERCEL_DETECTION = 
  process.env.USE_VERCEL_DETECTION?.toLowerCase() === "true" && 
  APP_ENV === "production";

/**
 * Environment flags for controlling deployment-specific behavior.
 * 
 * Set APP_ENV=development to enable localhost development mode:
 *   - Uses local storage instead of Supabase
 *   - Uses development cookies (not secure)
 *   - Uses localhost URLs instead of production URLs
 *   - Disables Railway/Vercel-specific detection
 * 
 * Set APP_ENV=production to enable production mode:
 *   - Uses Supabase storage
 *   - Uses production cookies (secure)
 *   - Uses production URLs
 *   - Enables Railway/Vercel-specific detection
 */
export const EnvFlags = {
  /** Current application environment */
  APP_ENV,
  
  /** Use local filesystem storage instead of Supabase Storage */
  USE_LOCAL_STORAGE,
  
  /** Use Supabase Storage for file uploads */
  USE_SUPABASE_STORAGE,
  
  /** Use production cookie policies (secure, SameSite=None) */
  USE_PRODUCTION_COOKIES,
  
  /** Use production URLs instead of localhost URLs */
  USE_PRODUCTION_URLS,
  
  /** Enable Railway-specific environment detection */
  USE_RAILWAY_DETECTION,
  
  /** Enable Vercel-specific environment detection */
  USE_VERCEL_DETECTION,
  
  /** Is running in development mode */
  IS_DEVELOPMENT: APP_ENV === "development",
  
  /** Is running in production mode */
  IS_PRODUCTION: APP_ENV === "production",
};

function isLocalUrl(url: string): boolean {
  const host = url.replace(/^https?:\/\//, "").split(/[/:]/)[0].toLowerCase()
  return host === "localhost" || host === "127.0.0.1" || host === "0.0.0.0"
}

/**
 * Get the appropriate API base URL based on environment.
 *
 * KEY: Use the Next.js proxy (/api) when in development — this is
 * **same-origin**, so no CORS preflight is needed at all.
 *
 * Resolution order:
 * 1. NEXT_PUBLIC_API_BASE_URL pointing to a remote host → use it (production)
 * 2. NEXT_PUBLIC_API_BASE_URL pointing to localhost → ignore, use proxy (dev)
 * 3. Not set → /api proxy (development default)
 */
export function getApiBaseUrl(): string {
  const configured = process.env.NEXT_PUBLIC_API_BASE_URL?.trim();
  if (configured) {
    // If it's a remote URL (not localhost), use it directly (production mode)
    if (!isLocalUrl(configured)) {
      return configured.endsWith("/api")
        ? configured.replace(/\/$/, "")
        : `${configured.replace(/\/$/, "")}/api`;
    }
  }
  // Development: use Next.js same-origin proxy — no CORS
  return "/api";
}

/**
 * Get the appropriate API origin (base URL without /api) for image resolution.
 * In development with proxy, returns the Flask server URL directly.
 */
export function getApiOrigin(): string {
  const base = getApiBaseUrl();
  // When using the Next.js proxy (relative URL), resolve to Flask server
  if (base === "/api") {
    return "http://127.0.0.1:5000";
  }
  return base.replace(/\/api(?:\/)?$/, "");
}